File: app/src/main/java/me/ykrank/s1next/view/fragment/FavouriteListPagerFragment.java
package me.ykrank.s1next.view.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import io.reactivex.Observable;
import me.ykrank.s1next.data.api.model.collection.Favourites;
import me.ykrank.s1next.data.api.model.wrapper.BaseResultWrapper;
import me.ykrank.s1next.util.L;
import me.ykrank.s1next.util.MathUtil;
import me.ykrank.s1next.view.adapter.FavouriteRecyclerViewAdapter;
import me.ykrank.s1next.view.internal.PagerCallback;
import me.ykrank.s1next.viewmodel.LoadingViewModel;

/**
 * A Fragment representing one of the pages of favourites.
 * <p>
 * Activity or Fragment containing this must implement {@link PagerCallback}.
 */
public final class FavouriteListPagerFragment extends BaseRecyclerViewFragment<BaseResultWrapper<Favourites>> {

    private static final String ARG_PAGE_NUM = "page_num";

    private int mPageNum;

    private FavouriteRecyclerViewAdapter mRecyclerAdapter;

    private PagerCallback mPagerCallback;

    public static FavouriteListPagerFragment newInstance(int pageNum) {
        FavouriteListPagerFragment fragment = new FavouriteListPagerFragment();
        Bundle bundle = new Bundle();
        bundle.putInt(ARG_PAGE_NUM, pageNum);
        fragment.setArguments(bundle);

        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        mPagerCallback = (PagerCallback) getFragmentManager().findFragmentByTag(
                FavouriteListFragment.Companion.getTAG());
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mPageNum = getArguments().getInt(ARG_PAGE_NUM);
        L.leaveMsg("FavouriteListPagerFragment##mPageNum" + mPageNum);

        RecyclerView recyclerView = getRecyclerView();
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mRecyclerAdapter = new FavouriteRecyclerViewAdapter(getActivity());
        recyclerView.setAdapter(mRecyclerAdapter);
    }

    @Override
    public void onDetach() {
        super.onDetach();

        mPagerCallback = null;
    }

    @Override
    Observable<BaseResultWrapper<Favourites>> getSourceObservable(@LoadingViewModel.LoadingDef int loading) {
        return mS1Service.getFavouritesWrapper(mPageNum);
    }

    @Override
    void onNext(BaseResultWrapper<Favourites> data) {
        Favourites favourites = data.getData();
        if (favourites.getFavouriteList() == null) {
            consumeResult(data.getResult());
        } else {
            super.onNext(data);

            mRecyclerAdapter.diffNewDataSet(favourites.getFavouriteList(), true);

            // update total page
            mPagerCallback.setTotalPages(MathUtil.divide(favourites.getTotal(),
                    favourites.getFavouritesPerPage()));
        }
    }
}

File: app/src/main/java/me/ykrank/s1next/view/fragment/FavouriteListPagerFragment.kt
package me.ykrank.s1next.view.fragment

import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import android.view.View
import com.github.ykrank.androidtools.ui.internal.PagerCallback
import com.github.ykrank.androidtools.ui.vm.LoadingViewModel
import com.github.ykrank.androidtools.util.MathUtil
import io.reactivex.Single
import me.ykrank.s1next.data.api.model.collection.Favourites
import me.ykrank.s1next.data.api.model.wrapper.BaseResultWrapper
import me.ykrank.s1next.view.adapter.FavouriteRecyclerViewAdapter

/**
 * A Fragment representing one of the pages of favourites.
 *
 *
 * Activity or Fragment containing this must implement [PagerCallback].
 */
class FavouriteListPagerFragment : BaseRecyclerViewFragment<BaseResultWrapper<Favourites>>() {

    private var mPageNum: Int = 0

    private lateinit var mRecyclerAdapter: FavouriteRecyclerViewAdapter

    private var mPagerCallback: PagerCallback? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPagerCallback = parentFragment as PagerCallback
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        mPageNum = arguments?.getInt(ARG_PAGE_NUM) ?: 0
        leavePageMsg("FavouriteListPagerFragment##mPageNum$mPageNum")

        val recyclerView = recyclerView
        recyclerView.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(context)
        mRecyclerAdapter = FavouriteRecyclerViewAdapter(activity)
        recyclerView.adapter = mRecyclerAdapter
    }

    override fun onDestroy() {
        mPagerCallback = null
        super.onDestroy()
    }

    override fun getSourceObservable(@LoadingViewModel.LoadingDef loading: Int): Single<BaseResultWrapper<Favourites>> {
        return mS1Service.getFavouritesWrapper(mPageNum)
    }

    override fun onNext(data: BaseResultWrapper<Favourites>) {
        val favourites = data.data
        if (favourites.favouriteList == null) {
            consumeResult(data.result)
        } else {
            super.onNext(data)

            mRecyclerAdapter.diffNewDataSet(favourites.favouriteList, true)

            // update total page
            mPagerCallback?.setTotalPages(MathUtil.divide(favourites.total,
                    favourites.favouritesPerPage))
        }
    }

    companion object {

        private const val ARG_PAGE_NUM = "page_num"

        fun newInstance(pageNum: Int): FavouriteListPagerFragment {
            val fragment = FavouriteListPagerFragment()
            val bundle = Bundle()
            bundle.putInt(ARG_PAGE_NUM, pageNum)
            fragment.arguments = bundle

            return fragment
        }
    }
}
